README
IIPImage - High Resolution Streaming Image Server
=================================================

[![GPLv3 Licensed](https://img.shields.io/badge/license-GPLv3-blue.svg)](https://www.gnu.org/copyleft/gpl.html)
[![Build Status](https://github.com/ruven/iipsrv/actions/workflows/c-cpp.yml/badge.svg?branch=master)](https://github.com/ruven/iipsrv/actions/workflows/c-cpp.yml)
[![Build Status](https://dev.azure.com/iipimage/iipsrv/_apis/build/status/iipsrv?branchName=master)](https://dev.azure.com/iipimage/iipsrv/_build/latest?definitionId=2&branchName=master)
[![CodeQL](https://github.com/ruven/iipsrv/actions/workflows/codeql-analysis.yml/badge.svg)](https://github.com/ruven/iipsrv/actions/workflows/codeql-analysis.yml)
[![Coverity Scan](https://img.shields.io/coverity/scan/3997.svg)](https://scan.coverity.com/projects/ruven-iipsrv)
[![CodeFactor](https://www.codefactor.io/repository/github/ruven/iipsrv/badge/master)](https://www.codefactor.io/repository/github/ruven/iipsrv/overview/master)

ABOUT
-----
IIPImage is an advanced high-performance feature-rich image server system for web-based streamed viewing and zooming of ultra high-resolution 
images. It is designed to be fast and bandwidth-efficient with low processor and memory requirements. The system can comfortably handle gigapixel size images as 
well as advanced image features such as 8, 16 and 32 bits per channel, CIELAB colorimetric images and scientific imagery such as multispectral images, image sequences and 3D surface topologies.


FEATURES
--------
* Fast lightweight FastCGI server
* High performance with inbuilt configurable cache
* Support for extremely large images: tera-pixel and multi gigapixel support
* TIFF and JPEG2000 input support
* JPEG, PNG and WebP output support: export of whole or regions of images at any size
* Supports IIP, Zoomify, DeepZoom and IIIF API's
* 1, 8, 16 and 32 bit image support including 32 bit floating point support
* CIELAB support with automatic CIELAB->sRGB colour space conversion
* Multispectral and hyperspectral image support
* Image stacks and image sequence support
* Dynamic hillshading of 3D surface topologies
* Dynamic watermarking
* Memcached support


DOCUMENTATION
-------------
Detailed class descriptions (generated using doxygen) are available in the
doc subdirectory



LICENCE
-------

iipsrv is released under the GNU General Public License (GPL). See the copyright
notice COPYING in this directory for licensing details or go to
http://www.gnu.org/licenses/gpl.html for more details.

If you use IIPImage on a public site and remove the IIP link logo from the 
client, you must provide a link on your site back to the IIPImage site - 
http://iipimage.sf.net

This distribution includes version 2.4.0 of the FCGI development libraries.
See COPYING.FCGI for licensing information for these libraries.



REQUIREMENTS
------------
Requirements: libtiff, zlib and the IJG JPEG development libraries.
Optional: libmemcached (for Memcached), libpng( for PNG output), libwebp (for WebP output) and Kakadu or OpenJPEG (for JPEG2000 input).

Plus, of course, an fcgi-enabled web server. The server has been successfully
tested on the following servers:
- [Apache](https://httpd.apache.org)
- [Lighttpd](https://www.lighttpd.net)
- [IIS](https://www.iis.net)
- [NginX](https://nginx.org)
- [MyServer](http://www.myserverproject.net)
- Java Application Servers (Tomcat, JBoss)

Example server configurations are shown below.



BUILDING
--------
The standard autoconf build process should work fine. If you want to allow
dynamic loading of 3rd party image decoders, use the configure option
--enable-modules. There is a version of the FCGI development library included
in this distribution. The configure script will use this bundled version
unless it detects one already installed. Alternatively, you may specify the
path using --with-fcgi-incl=<path> and --with-fcgi-lib=<path>.

If this is an SVN or Git development version, first generate the autoconfigure
environment using autogen.sh:

    ./autogen.sh

Otherwise for release versions, use configure directly:

    ./configure
    make


OPTIONAL LIBRARIES: MEMCACHED
-----------------------------
IIPImage is able to use Memcached (http://www.memcached.org), a high-performance,
distributed memory object caching system. If enabled, IIPImage will cache
results using Memcached, giving IIPImage added speed and scalability. To use
this, you will need to install the library (and development files) of
libmemcached (http://libmemcached.org). This will be automatically detected
during the build process. 



OPTIONAL LIBRARIES: KAKADU
--------------------------
IIPImage is able to decode JPEG2000 images via the Kakadu SDK
(http://www.kakadusoftware.com). This is, however, not open source and you will
need to purchase a license for the source code. In order to use, first build
the Kakadu SDK as per the instructions supplied with the SDK. Then, supply the
following parameters to the ./configure command

    --with-kakadu=/path/to/kakadu/distribution



OPTIONAL LIBRARIES: OPENJPEG
----------------------------

IIPImage is able to decode JPEG2000 images via the open source OpenJPEG 
library (http://www.openjpeg.org/). Activated if OpenJPEG development
files found in system path or path provided via

    --with-openjpeg=/path/to/openjpeg/distribution

OpenJPEG will be disabled if Kakadu support has been requested or --disable-openjpeg used.



INSTALLATION
------------
Simply copy the executable called iipsrv.fcgi in the src subdirectory into
the web server fcgi directory. If one does not exist, simply create one, called,
for example, fcgi-bin. The web server will need to be configured to use this
executable.



CONFIGURATION
-------------
There are several startup variables that can be passed to the server.
They are all optional.

LOGFILE: the server will log its output to the file specified, if it can.
Version 1.1 and later of iipsrv also supports logging to syslog if “syslog”
is given as the value.

VERBOSITY: 0 means no logging, 1 is minimal logging, 2 lots of debugging stuff,
3 even more debugging stuff and 10 a very large amount indeed ;-)

MAX_IMAGE_CACHE_SIZE: Max image cache size to be held in RAM in MB. This is
a cache of the compressed JPEG image tiles requested by the client.
The default is 10MB.

FILESYSTEM_PREFIX: This is a prefix automatically added by the server to the 
beginning of each file system path. This can be useful for security reasons to 
limit access to certain sub-directories. For example, with a prefix of 
"/home/images/" set on the server, a request by a client for "image.tif" will 
point to the path "/home/images/image.tif".  Any reverse directory path 
component such as ../ is also filtered out. No default value.

FILESYSTEM_SUFFIX: This  is a suffix added to the end of each file system path.
It can be combined with FILESYSTEM_PREFIX. It is not used in combination with
FILENAME_PATTERN. If e.g. this is set to ".tif", an image URL such as  "/UUID"
will look for "${FILESYSTEM_PREFIX}/UUID.tif". In the IIIF info.json document,
the image @id will be set without the ".tif" suffix.

JPEG_QUALITY: The default JPEG quality factor for compression when the
client does not specify one. The value should be between 1 (highest level of
compression) and 100 (highest image quality). The default is 75.

PNG_QUALITY: The default PNG quality factor for compression when the
client does not specify one. The value should be between 1 (highest level of
compression) and 9 (highest image quality). The default is 1.

WEBP_QUALITY: The default WebP quality factor for compression when the
client does not specify one. The value should be between 0 (highest level of
compression) and 100 (highest image quality). The default is 50.

MAX_CVT: Limits the maximum output image dimensions (in pixels) allowable for dynamic
image export via the CVT command or for IIIF requests. This prevents huge requests
from overloading the server. The default is 5000. If set to -1, no limit is set.

ALLOW_UPSCALING: Determines whether an image may be rendered at a size greater
than that of the source image. A value of 0 will prevent upscaling.
The default is 1 (upscaling is allowed).

MAX_LAYERS: The maximum number of quality layers to decode for images that support 
progressive quality encoding, such as JPEG2000. Ignored for other file 
formats. If not set, half of the available quality layers will be decoded by default.
If set to -1, all the available layers will be decoded by default.

FILENAME_PATTERN: Pattern that follows the name stem for a 3D or multispectral 
sequence. eg: "_pyr_" for FZ1_pyr_000_090.tif. The default is "_pyr_". This is 
only relevant to 3D image sequences.

WATERMARK: TIFF image to use as watermark file. This image should be not be 
bigger the tile size used for TIFF tiling. If bigger, it will simply be 
cropped to the tile size. If smaller, the watermark will be positioned 
randomly within the available space. The image can be either colour or 
grayscale.

WATERMARK_PROBABILITY: The probability that a particular tile will have a watermark
applied to it. 0 means never, 1 means always.

WATERMARK_OPACITY: The opacity (between 0 and 1) applied to the watermark image.

MEMCACHED_SERVERS: A comma-delimitted list of memcached servers with optional
port numbers. For example: localhost,192.168.0.1:8888,192.168.0.2.

MEMCACHED_TIMEOUT: Time in seconds that cache remains fresh.
Default is 86400 seconds (24 hours).

INTERPOLATION: Interpolation method to use for rescaling when using image export.
Integer value. 0 for fastest nearest neighbour interpolation. 1 for bilinear
interpolation (better quality but about 2.5x slower). Bilinear by default.

CORS: Cross Origin Resource Sharing setting. Disabled by default.
Set to * to enable for all domains or specify a single domain.
See http://www.w3.org/TR/cors/ for more details on CORS.

BASE_URL: Set a base URL for use in certain protocol requests if web server rewriting 
has taken place and the public URL is not the same as that supplied to iipsrv.

URI_MAP: Set a mapping from a URL prefix to a supported protocol. This enables iipsrv to
be able to work without requiring full CGI query strings. Map must be of the form
"prefix=>protocol" where prefix can be either empty or any string prefix and protocol must
be one of IIP,IIIF,DeepZoom, Zoomify. Used, for example, to map requests of the form
http://server/iiif/ to the IIIF protocol handler without requiring web server rewriting.

CACHE_CONTROL: Set the HTTP Cache-Control header. See http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.9 for 
a full list of options. If not set, header defaults to "max-age=86400" (24 hours).

EMBED_ICC: Set whether the ICC profile is embedded within the output image.
0 to strip profile, 1 to embed profile. The default is 1 (embedded profiles).

OMP_NUM_THREADS: Set the number of OpenMP threads to be used by the iipsrv image
processing routines (See OpenMP specification for details). All available processor
threads are used by default.

KAKADU_READMODE: Set the Kakadu JPEG2000 read-mode. 0 for 'fast' mode with minimal error checking (default), 1 for 'fussy' mode with no error 
recovery, 2 for 'resilient' mode with maximum recovery from codestream errors. See the Kakadu documentation for further details.

IIIF_VERSION: Set the major IIIF Image API version. Values should be a single digit. For example: 2 for versions 2 or 2.1 etc.
3 for IIIF version 3.x. If not set, defaults to version IIIF 3.x

DECODER_MODULES: Comma separated list of external modules for decoding 
other image formats. This is only necessary if you have activated 
--enable-modules for ./configure and written your own image format 
handler(s).




IMAGE PATHS
-----------
The images paths given to the server via the FIF variable must be
absolute paths on the server machine (eg. FIF=/images/test.tif) 
and not paths relative to the web server document root location.
Images do not, therefore, need to be directly accessible by the
client via the web server. Make sure the server process owner is
able to access and read the images!



EXAMPLE SERVER CONFIGURATIONS
-----------------------------

### Apache and mod_fastcgi

httpd.conf example extract:

```
--------------------------------------------------------------------
# Create a directory for the iipsrv binary
ScriptAlias /fcgi-bin/ "/usr/local/httpd/fcgi-bin/"

# Set the options on that directory
<Directory "/usr/local/httpd/fcgi-bin">
   AllowOverride None
   Options None

# Syntax for access is different in Apache 2.4 - uncomment appropriate version
# Apache 2.2
#   Order allow,deny
#   Allow from all

# Apache 2.4
   Require all granted

</Directory>

# Set the module handler
AddHandler fastcgi-script fcg fcgi fpl

# Initialise some variables for the FCGI server
FastCgiServer /usr/local/httpd/fcgi-bin/iipsrv.fcgi \
-initial-env LOGFILE=/tmp/iipsrv.log \
-initial-env VERBOSITY=2 \
-initial-env MAX_IMAGE_CACHE_SIZE=10 \
-initial-env FILENAME_PATTERN=_pyr_ \
-initial-env JPEG_QUALITY=50 \
-initial-env MAX_CVT=3000 \
-listen-queue-depth 2048 \
-processes 1
--------------------------------------------------------------------
```



### Apache and mod_fcgid

mod_fcgid is a binary compatible replacement for mod_fastcgi. It works in the 
same way, but is configured differently. Load the module like this:


    LoadModule fcgid_module /path/to/apachemodules/mod_fcgid.so


Here is an example configuration. Note that mod_fcgid does not have a 
FastCgiServer directive and there is no need to explicitly start the server:

```
-------------------------------------------------------------------
# Create a directory for the iipsrv binary
ScriptAlias /fcgi-bin/ "/var/www/localhost/fcgi-bin/"

# Set the options on that directory
<Directory "/var/www/localhost/fcgi-bin/">
   AllowOverride None
   Options None

# Syntax for access is different in Apache 2.4 - uncomment appropriate version
# Apache 2.2
#   Order allow,deny
#   Allow from all

# Apache 2.4
   Require all granted

   # Set the module handler
   AddHandler fcgid-script .fcgi
</Directory>

# Set our environment variables for the IIP server
FcgidInitialEnv VERBOSITY "5"
FcgidInitialEnv LOGFILE "/tmp/iipsrv.log"
FcgidInitialEnv MAX_IMAGE_CACHE_SIZE "10"
FcgidInitialEnv JPEG_QUALITY "50"
FcgidInitialEnv MAX_CVT "3000"

# Define the idle timeout as unlimited and the number of
# processes we want
FcgidIdleTimeout 0
FcgidMaxProcessesPerClass 1

-------------------------------------------------------------------
```

Note that on CentOS, FcgidIPCDir is configured by default to 
/var/log/httpd/fcgidsock, which may not be writable by Apache. If this is the case, 
specify another location for FcgidIPCDir, which is writable, such as /tmp/fcgidsock


### Lighttpd
lighttpd.conf example extract:

```
--------------------------------------------------------------------
fastcgi.server = ( "/fcgi-bin/iipsrv.fcgi" =>
  (( "host" => "127.0.0.1",
     "port" => 9000,
     "check-local" => "disable",
     "min-procs" => 1,
     "max-procs" => 1,
     "bin-path" => "/var/www/localhost/fcgi-bin/iipsrv.fcgi",
     "bin-environment" => (
        "LOGFILE" => "/tmp/iipsrv.log",
        "VERBOSITY" => "5",
        "MAX_IMAGE_CACHE_SIZE" => "10",
        "FILENAME_PATTERN" => "_pyr_",
        "JPEG_QUALITY" => "50",
        "MAX_CVT" => "3000"
      )
  ))
)
--------------------------------------------------------------------
```

### NginX

iipsrv can also be used with NginX. The one drawback with regards to iipsrv is that it cannot automatically spawn FCGI processes, so you will need to use 
nginx together with spawn-fcgi or start iipsrv from the command line.

To set up nginx with iipsrv add a directive such as this to your nginx configuration, which will forward requests to /fcgi-bin/iipsrv.fcgi to a 
running iipsrv process on port 9000.

```
location /fcgi-bin/iipsrv.fcgi {
	fastcgi_pass    localhost:9000;
	fastcgi_param   PATH_INFO $fastcgi_script_name;
        fastcgi_param   REQUEST_METHOD $request_method;
        fastcgi_param   QUERY_STRING $query_string;
        fastcgi_param   CONTENT_TYPE $content_type;
        fastcgi_param   CONTENT_LENGTH $content_length;
        fastcgi_param   SERVER_PROTOCOL $server_protocol;
        fastcgi_param   REQUEST_URI $request_uri;
        fastcgi_param   HTTPS $https if_not_empty;
}
```

Nginx can also handle load balancing to multiple iipsrv instances, which can be hosted on the same machine on multiple ports or on different hosts. 
For a multiple host configuration, declare the load balancing like this:

```
upstream iip {
	server 192.168.0.1:9000;
	server 192.168.0.2:9000;
	server 192.168.0.3:9000;
	server 192.168.0.4:9000;
}
```

and change the fastcgi_pass parameter in the above location configuration to point to this instead of a fixed address:

    fastcgi_pass       iip;


### spawn-fcgi

iipsrv can also be used with lighttpd's spawn-fcgi without the need for a 
full web server. Simply spawn the iipsrv process on the command line. The 
process can be bound to an IP address and port for backend load-balancing 
configurations. For example:

    spawn-fcgi -f src/iipsrv.fcgi -a 192.168.0.1 -p 9000


### MyServer:
Simply run the MyServer configuration and in the MIME section, choose the .fcgi extension
and select:
```
--------------------------------------------------------------------
MIME Type: application octet-stream
Action: Execute self contained FastCGI
Manager: NONE 
--------------------------------------------------------------------
```


### Java Application Servers (Tomcat, Jetty, JBoss etc)

IIPImage can also be used with Java Application Servers such as Apache Tomcat, JBoss and Jetty. Simply add 
the JFastCGI jar file to your webapp and add the following to your web.xml configuration file in order to 
re-route FCGI requests to the IIPImage server on the specified port.

```
<!-- Gateway Servlet to IIPImage FCGI server -->

<servlet>
  <servlet-name>fcgi</servlet-name>
  <servlet-class>net.jr.fastcgi.FastCGIServlet</servlet-class>
  <init-param>>
    <param-name>server-address</param-name>
   <param-value>127.0.0.1:6667</param-value>
  </init-param>
</servlet>

<servlet-mapping>
  <servlet-name>fcgi</servlet-name>
  <url-pattern>/cgi-bin/iipsrv.fcgi</url-pattern>
</servlet-mapping>
```

You then need to start an instance of the server on the requested port (6667 in this example) using 
spawn-cgi (see spawn-cgi section above) or on the command line (see below).


### Command Line

It is also possible to start iipsrv directly on the command line using the --bind parameter.
For example:

    iipsrv.fcgi --bind 192.168.0.1:9000

where the argument given to bind is the socket on which to listen to FCGI
requests. Note that these are *not* HTTP requests and iipsrv will still require a web server front-end.

There is additionally a --backlog parameter that is optional and sets the socket backlog value. The backlog value specifies the number of requests can be queued and, 
therefore, increases the number of concurrent connections that iipsrv can handle and is set to 2048 by default. For example:

    iipsrv.fcgi --bind 192.168.0.1:9000 --backlog 1024

Note that the backlog parameter must be specified after the bind parameter and argument.  Note also that this value may be limited by the operating system. On Linux kernels 
< 2.4.25 and Mac OS X, the backlog limit is hard-coded to 128, so any value above this will be limited to 128 by the OS. If you do provide a backlog value, verify whether 
the setting /proc/sys/net/core/somaxconn should be updated.

Your web server should, therefore, be configured to use this address for FastCGI.
For example with lighttpd:

    fastcgi.server = (
      "/fcgi-bin/iipsrv.fcgi" => (
            ("host"=>"192.168.0.1", "port"=>9000, "check-local"=>"disable")
      )
    )

For Apache, load mod_proxy and mod_proxy_fcgi and add the following line to Apache's configuration:

    ProxyPass "/fcgi-bin/iipsrv.fcgi" "fcgi://192.168.0.1:9000/"



DEBUGGING
---------

Basic debugging can be carried out by simply using the server logs. Verbose logging can be obtained
by increasing the VERBOSITY startup variable. Implemented levels of verbosity range from 0 (no logging)
to 6 (maximum amount of logging). The pre-processor defines KAKADU_DEBUG and OPENJPEG_DEBUG can
additionally be set at compile time to generate additional detailed logging from these particular decoders.

It is also possible to test iipsrv without the use of FCGI or a web server. In this case, you will need to
compile iipsrv with DEBUG defined for the pre-processor. This can be done during the ./configure stage before
building using make:

    CXXFLAGS="-DDEBUG" ./configure
    make

The resulting binary can be started on the command line and will accept any supported imaging protocol requests
(IIP, IIIF, Deepzoom or Zoomify) without any HTTP prefix or domain name (e.g. FIF=test.tif&WID=500&CVT=JPEG) as
standard input on the command line and output the image or data to the file "iipsrv.debug" in the same directory.
Successive requests will replace the contents of iipsrv.debug with the output of the new request. For example:

    % ./src/iipsrv.fcgi
    FIF=test.tif&WID=500&CVT=JPEG
    IIIF=/test.tif/full/500,/0/default.jpg
    FIF=test.tif&WID=600&ROT=90&&CVT=PNG

On Linux/Unix based systems, it's also possible to use pipes or redirection to send a list of requests within a file
with iipsrv automatically terminating at the end of the list:

    iipsrv.fcgi < request_list.txt

Debugging using memory testing tools, such as valgrind (https://valgrind.org/), can be performed by starting
iipsrv on the command-line through the testing tool. For example:

    valgrind iipsrv.fcgi --bind 0.0.0.0:9000

This can work in conjunction with debugging mode or together with a web server in normal FCGI-mode. In this case
the web server must be setup to proxy forward FCGI requests to the address and port specified when starting iipsrv.



------------------------------------------------------------------------------------
Please refer to the project site https://iipimage.sourceforge.io for further details

------------------------------------------------------------------------------------
(c) 2000-2023 Ruven Pillay <ruven@users.sourceforge.net>


NEWS
See https://iipimage.sourceforge.io/blog for the latest IIPImage news


src/OBJ.cc
/*
    IIP OJB Command Handler Class Member Functions

    Copyright (C) 2006-2023 Ruven Pillay.

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software Foundation,
    Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
*/


#include "Task.h"
#include <algorithm>
#include <sstream>


using namespace std;



void OBJ::run( Session* s, const std::string& a )
{

  argument = a;
  // Convert to lower case the argument supplied to the OBJ command
  transform( argument.begin(), argument.end(), argument.begin(), ::tolower );

  session = s;

  // Log this
  if( session->loglevel >= 3 ) *(session->logfile) << "OBJ :: " << argument << " to be handled" << endl;

  // Time this command
  if( session->loglevel >= 2 ) command_timer.start();


  if( argument == "iip,1.0" ) iip();
  else if( argument == "basic-info" ){
    iip_server();
    max_size();
    resolution_number();
    colorspace( "*,*" );
  }
  else if( argument == "iip-server" ) iip_server();
  // IIP optional commands
  else if( argument == "iip-opt-comm" ) session->response->addResponse( "IIP-opt-comm:CVT CNT QLT JTL JTLS WID HEI RGN MINMAX SHD CMP INV CTW" );
  // IIP optional objects
  else if( argument == "iip-opt-obj" ) session->response->addResponse( "IIP-opt-obj:Horizontal-views Vertical-views Tile-size Bits-per-channel Min-Max-sample-values Resolutions" );
  // Resolution-number
  else if( argument == "resolution-number" ) resolution_number();
  // Max-size
  else if( argument == "max-size" ) max_size();
  // Tile-size
  else if( argument == "tile-size" ) tile_size();
  // Bits per pixel
  else if( argument == "bits-per-channel" ) bits_per_channel();
  // Vertical-views
  else if( argument == "vertical-views" ) vertical_views();
  // Horizontal-views
  else if( argument == "horizontal-views" ) horizontal_views();
  // Minimum and maximum provided by TIFF tags
  else if( argument == "min-max-sample-values" ) min_max_values();
  // List of available resolutions
  else if( argument == "resolutions" ) resolutions();
  // Get physical resolution (DPI)
  else if( argument == "dpi" ) dpi();
  else if( argument == "stack" ) stack();

  // Colorspace
  /* The request can have a suffix, which we don't need, so do a
     like scan
  */
  else if( argument.find( "colorspace" ) != string::npos ){
    colorspace( "*,*" );
  }

  // Image Metadata
  else if( argument == "summary-info" ){
    metadata( "rights" );
    metadata( "description" );
    metadata( "creator" );
    metadata( "date" );
    metadata( "software" );
  }

  else if( argument == "rights" || argument == "title" ||
	   argument == "description" || argument == "creator" ||
	   argument == "keywords" || argument == "comment" ||
	   argument == "last-author" || argument == "rev-number" ||
	   argument == "edit-time" || argument == "last-printed" ||
	   argument == "date" || argument == "last-save-dtm" ||
	   argument == "software"  || argument == "make" ||
	   argument == "model" || argument == "xmp" ||
	   argument == "scale" ){

    metadata( argument );
  }

  // Send all available metadata
  else if( argument == "metadata" ){

    stringstream json;
    json << "{ ";

    map <const string, string> metadata = (*session->image)->metadata;
    map<const string,string> :: const_iterator i;
    for( i = metadata.begin(); i != metadata.end(); i++ ){
      if( i->first == "icc" || i->first == "xmp" ) continue;
      if( (i->second).length() ) json << endl << "\t\"" << i->first << "\": \"" << i->second << "\",";
    }

    // Remove trailling comma - if no items were added, this safely removes extra white space after opening {
    json.seekp( -1, std::ios_base::end );
    json << endl << "}";

    session->response->setMimeType( "application/json" );
    session->response->addResponse( json.str() );
  }

  // None of the above!
  else{

    if( session->loglevel >= 1 ){
      *(session->logfile) << "OBJ :: Unsupported argument: " << argument << " received" << endl;
    }

    // Unsupported object error code is 3 2
    session->response->setError( "3 2", argument );
  }


  if( session->loglevel >= 2 ){
    *(session->logfile) << "OBJ :: Total command time " << command_timer.getTime() << " microseconds" << endl;
  }


}



void OBJ::iip(){
  session->response->setProtocol( "IIP:1.0" );
}


void OBJ::iip_server(){
  // The binary capability code is 1000001 == 65 in integer
  // ie can do CVT jpeg and JTL, but no transforms
  session->response->addResponse( "IIP-server:3.65" );
}


void OBJ::max_size(){
  checkImage();
  int x = (*session->image)->getImageWidth();
  int y = (*session->image)->getImageHeight();

  // For 90 and 270 rotation swap width and height
  if( (int)((session->view)->getRotation()) % 180 == 90 ){
    unsigned int tmp = x;
    x = y;
    y = tmp;
  }

  if( session->loglevel >= 2 ){
    *(session->logfile) << "OBJ :: Max-size is " << x << " " << y << endl;
  }
  session->response->addResponse( "Max-size", x, y );
}


void OBJ::resolution_number(){

  checkImage();
  int no_res = (*session->image)->getNumResolutions();
  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: Resolution-number handler returning " << no_res << endl;
  }
  session->response->addResponse( "Resolution-number", no_res );

}


void OBJ::dpi(){

  checkImage();
  float dpix = (*session->image)->getHorizontalDPI();
  float dpiy = (*session->image)->getVerticalDPI();

  if( dpix && dpiy ){
    char tmp[64];
    snprintf( tmp, 64, "DPI:%f %f", dpix, dpiy );
    if( session->loglevel >= 5 ){
      *(session->logfile) << "OBJ :: DPI handler returning " << tmp << endl;
    }
    session->response->addResponse( tmp );
  }
}


void OBJ::tile_size(){
  checkImage();

  int x = (*session->image)->getTileWidth();
  int y = (*session->image)->getTileHeight();
  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: Tile-size is " << x << " " << y << endl;
  }
  session->response->addResponse( "Tile-size", x, y );
}


void OBJ::bits_per_channel(){

  checkImage();
  int bpc = (*session->image)->getNumBitsPerPixel();
  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: Bits-per-channel handler returning " << bpc << endl;
  }
  session->response->addResponse( "Bits-per-channel", bpc );

}


void OBJ::vertical_views(){
  checkImage();
  list <int> views = (*session->image)->getVerticalViewsList();
  list <int> :: const_iterator i;
  string tmp = "Vertical-views:";
  char val[8];
  for( i = views.begin(); i != views.end(); i++ ){
    snprintf( val, 8, "%d ", *i );
    tmp += val;
  }
  // Chop off the final space
  tmp.resize( tmp.length() - 1 );
  session->response->addResponse( tmp );
}


void OBJ::horizontal_views(){
  checkImage();
  list <int> views = (*session->image)->getHorizontalViewsList();
  list <int> :: const_iterator i;
  string tmp = "Horizontal-views:";
  char val[8];
  for( i = views.begin(); i != views.end(); i++ ){
    snprintf( val, 8, "%d ", *i );
    tmp += val;
  }
  // Chop off the final space
  tmp.resize( tmp.length() - 1 );
  session->response->addResponse( tmp );
}


void OBJ::min_max_values(){

  checkImage();
  unsigned int n = (*session->image)->getNumChannels();
  string tmp = "Min-Max-sample-values:";
  char val[24];
  float minimum, maximum;
  for( unsigned int i=0; i<n ; i++ ){
    minimum = (*session->image)->getMinValue(i);
    maximum = (*session->image)->getMaxValue(i);
    snprintf( val, 24, " %.9g ", minimum );
    tmp += val;
    snprintf( val, 24, " %.9g ", maximum );
    tmp += val;
  }
  // Chop off the final space
  tmp.resize( tmp.length() - 1 );
  session->response->addResponse( tmp );
  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: Min-Max-sample-values handler returning " << tmp << endl;
  }

}


void OBJ::resolutions(){

  checkImage();
  char val[32];
  int num_res = (*session->image)->getNumResolutions();

  string tmp = "Resolutions:";
  for( int i=num_res-1; i>=0; i-- ){
    snprintf( val, 32, "%d %d", (*session->image)->image_widths[i], (*session->image)->image_heights[i] );
    tmp += val;
    if( i>0 ) tmp += ",";
  }
  session->response->addResponse( tmp );
  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: Resolutions handler returning " << tmp << endl;
  }
}


void OBJ::colorspace( std::string arg ){

  checkImage();

  /* Assign the colourspace tag: 1 for greyscale, 3 for RGB and
     a colourspace of 4 to LAB images
     WARNING: LAB support is an extension and is not in the
     IIP protocol standard (as of version 1.05)
  */
  const char *planes = "3 0 1 2";
  int calibrated = 0;
  int colourspace;
  if( (*session->image)->getColourSpace() == CIELAB ){
    colourspace = 4;
    calibrated = 1;
  }
  else if( (*session->image)->getColourSpace() == GREYSCALE ){
    colourspace = 1;
    planes = "1 0";
  }
  else colourspace = 3;

  int no_res = (*session->image)->getNumResolutions();
  char tmp[41];
  snprintf( tmp, 41, "Colorspace,0-%d,0:%d 0 %d %s", no_res-1,
	    calibrated, colourspace, planes );

  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: Colourspace handler returning " << tmp << endl;
  }

  session->response->addResponse( tmp );
}


void OBJ::metadata( string field ){

  checkImage();

  string metadata = (*session->image)->metadata[field];

  if( session->loglevel >= 5 ){
    *(session->logfile) << "OBJ :: " << field << " handler returning '" << metadata << "'" << endl;
  }

  if( metadata.length() ){
    // Set appropriate mime type
    string mimeType = "text/plain";
    if( field == "xmp" ) mimeType = "application/xml";
    session->response->setMimeType( mimeType );
    session->response->addResponse( metadata );
  }
}


// Return image stack metadata in JSON format
void OBJ::stack(){

  checkImage();

  if( (*session->image)->isStack() ){

    list<Stack> items = (*session->image)->getStack();
    list<Stack> :: const_iterator i;
    int n = 0;

    stringstream json;
    json.precision(9);
    json << "[ ";

    for( i=items.begin(); i != items.end(); i++ ){
      json << endl << "\t{" << endl
	   << "\t\t\"id\": " << n++ << "," << endl
	   << "\t\t\"name\": \"" << (*i).name << "\"," << endl
	   << "\t\t\"scale\": " << (*i).scale << endl << "\t},";
    }

    // Remove trailling comma - if no items were added, this safely removes extra white space after opening {
    json.seekp( -1, std::ios_base::end );
    json << endl << "]";

    session->response->setMimeType( "application/json" );
    session->response->addResponse( json.str() );
  }
  else{
    if( session->loglevel >= 3 ){
      *(session->logfile) << "OBJ :: stack handler: not an image stack" << endl;
    }
  }
}
